allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,320 @@
+        embedded_input = self._target_embedder(input_indices)
+        if self._attention_function:
+            # encoder_outputs : (batch_size, input_sequence_length, encoder_output_dim)
+            # Ensuring mask is also a FloatTensor. Or else the multiplication within attention will complain.
+            encoder_outputs_mask = encoder_outputs_mask.type(torch.FloatTensor)
+            # (batch_size, input_sequence_length)
+            input_weights = self._decoder_attention(decoder_hidden_state, encoder_outputs, encoder_outputs_mask)
+            # (batch_size, encoder_output_dim)
+            attended_input = weighted_sum(encoder_outputs, input_weights)
+            # (batch_size, encoder_output_dim + target_embedding_dim)
+            return torch.cat((attended_input, embedded_input), -1)
+        else:
+            return embedded_input
+
+    def _predict_step(self, # pylint: disable=no-self-use
+                      target_scores: torch.LongTensor) -> Dict[str, torch.LongTensor]:
+        """
+        Take the projected scores for all output classes and predict the output for a single timestep.
+        Here we simply normalize the scores and output the class with the highest probability as the
+        prediction. This method can be overridden to do a constrained prediction and/or avoid local
+        normalization.
+
+        Parameters
+        ----------
+        target_scores : torch.LongTensor
+            Unnormalized class scores, the output from a projection layer. (batch_size, num_classes)
+
+        Returns
+        -------
+        prediction_dict : Dict[str, torch.LongTensor] with fields
+            class_probabilities
+            predicted_classes
+        """
+        # (batch_size, num_classes)
+        class_probabilities = F.softmax(target_scores)
+        # (batch_size,)
+        _, predicted_classes = torch.max(class_probabilities, 1)
+        prediction_dict = {"class_probabilities": class_probabilities, "predicted_classes": predicted_classes}
+        return prediction_dict
+
+    @staticmethod
+    def _get_loss(logits: torch.LongTensor,
+                  targets: torch.LongTensor,
+                  target_mask: torch.LongTensor) -> torch.LongTensor:
+        """
+        Takes logits (unnormalized outputs from the decoder) of size (batch_size, num_decoding_steps - 1,
+        num_classes), target indices of size (batch_size, num_decoding_steps) and corresponding masks of size
+        (batch_size, num_decoding) steps and computes cross entropy loss while taking the mask into account.
+
+        The length of ``targets`` is expected to be greater than that of ``logits`` because the decoder does
+        not need to compute the output corresponding to the last timestep of ``targets``. This method aligns
+        the inputs appropriately to compute the loss.
+        """
+        # `targets` currently contains START + target_sequence + END, and logits correspond to the outputs
```

Comment:
I'd probably integrate this comment into the docstring, now that you have a separate method for this.  You've got some duplication between the docstring and the comment.